### PR TITLE
xquartz: fix build

### LINKED ISSUE #40196: XQuartz fails to build
## Issue description

mesa-noglu, which doesn't support darwin, is a dependency for xorg-server, which is in turn a dependency for xquartz (which only makes sense for darwin AFAIK), breaking the build.

### Steps to reproduce

```
 nix-env --show-trace -iA nixpkgs.xquartz
```

should result in:

```
installing 'xquartz'
error: while evaluating the attribute 'installPhase' of the derivation 'xquartz' at /nix/store/np9vzblwz58gprh62rl0mi6166gb04sf-nixpkgs-18.09pre139198.9d0b6b9dfc9/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute 'configureFlags' of the derivation 'xinit-1.4.0' at /nix/store/np9vzblwz58gprh62rl0mi6166gb04sf-nixpkgs-18.09pre139198.9d0b6b9dfc9/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute 'postInstall' of the derivation 'xorg-server-1.18.4' at /nix/store/np9vzblwz58gprh62rl0mi6166gb04sf-nixpkgs-18.09pre139198.9d0b6b9dfc9/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute 'propagatedBuildInputs' of the derivation 'xorg-server-1.18.4' at /nix/store/np9vzblwz58gprh62rl0mi6166gb04sf-nixpkgs-18.09pre139198.9d0b6b9dfc9/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute 'handled' at /nix/store/np9vzblwz58gprh62rl0mi6166gb04sf-nixpkgs-18.09pre139198.9d0b6b9dfc9/nixpkgs/pkgs/stdenv/generic/check-meta.nix:213:7:
while evaluating 'handleEvalIssue' at /nix/store/np9vzblwz58gprh62rl0mi6166gb04sf-nixpkgs-18.09pre139198.9d0b6b9dfc9/nixpkgs/pkgs/stdenv/generic/check-meta.nix:128:28, called from /nix/store/np9vzblwz58gprh62rl0mi6166gb04sf-nixpkgs-18.09pre139198.9d0b6b9dfc9/nixpkgs/pkgs/stdenv/generic/check-meta.nix:214:14:
Package ‘mesa-noglu-17.3.8’ in /nix/store/np9vzblwz58gprh62rl0mi6166gb04sf-nixpkgs-18.09pre139198.9d0b6b9dfc9/nixpkgs/pkgs/development/libraries/mesa/default.nix:266 is not supported on ‘x86_64-apple-darwin’, refusing to evaluate.

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.
```

## Technical details

 - system: `"x86_64-darwin"`
 - host os: `Darwin 17.5.0, macOS 10.13.4`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0.2`
 - channels(delta): `"darwin"`
 - channels(root): `"nixpkgs-18.09pre139198.9d0b6b9dfc9"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`

### PR DESCRIPTION
**Pull request by vcunat**
Things done:

- use libGLU instead of mesa for darwin support
- move patches from local to github url
- fixup xquartz install

There may still be some issues at runtime. PRs welcome!

Fixes #40196

(cherry picked from commit c839771129f17cad501effb6676b0e0cde5cbd0b)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by vcunat** [comment]
@GrahamcOfBorg build xorg.xorgserver

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -12864,12 +12864,11 @@ with pkgs;
     inherit clangStdenv fetchurl fetchgit fetchpatch stdenv intltool freetype fontconfig
       libxslt expat libpng zlib perl mesa_drivers spice-protocol libunwind
       dbus libuuid openssl gperf m4 libevdev tradcpp libinput mcpp makeWrapper autoreconfHook
-      autoconf automake libtool mtdev pixman
+      autoconf automake libtool mtdev pixman libGL libGLU
       cairo epoxy;
     inherit (buildPackages) pkgconfig xmlto asciidoc flex bison;
     inherit (darwin) apple_sdk cf-private libobjc;
     bootstrap_cmds = if stdenv.isDarwin then darwin.bootstrap_cmds else null;
-    mesa = libGL;
```
Comment (xeji) [inline pkgs/top-level/all-packages.nix]: looks like this is still needed somewhere (eval fail)

**Comment by matthewbauer** [comment]
Does this work as expected for you? I was a little concerned that there might still be some patches needed.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (vcunat) [reply pkgs/top-level/all-packages.nix]: I know and have a fix; I'm just looking at other things in the meantime.

Comment (matthewbauer) [reply pkgs/top-level/all-packages.nix]: Yes i think abbradar’s commit is needed to cherrypick.

**Comment by vcunat** [comment]
On current 18.03 it doesn't even build.  I don't have a Darwin, and that's why I created the PR to use the Borg.

**Comment by vcunat** [comment]
@matthewbauer if you (or someone else) can test this on 18.03, that would be great.  Even if not and borg confirms it builds, I'd merge it.  I'm confident non-Darwin platforms won't get broken.

**Comment by matthewbauer** [comment]
@vcunat can you check out this failure on Hydra:

https://hydra.nixos.org/build/74047022

It might be caused by my commit, but I honestly have no idea what is going on there. It could be unrelated but I'm not sure if anyone else has been messing with xorg stuff.

**Comment by matthewbauer** [comment]
So, it looks like it might have to do with the s/mesa_noglu/libGLU/ change. I misread the original as ```mesa_glu``` and ```libGLU = libGLDarwinOr mesa_glu```. I believe to work correctly you would have to do something like this as a build input:

```
(if stdenv.isDarwin then libGLDarwin else mesa_noglu)
```

BTW I find the ```libGLDarwinOr``` incredibly confusing to work with. I don't have any time right now but if anyone can, a commit replacing all ```libGLDarwinOr``` with ```if stdenv.isDarwin then libGLDarwin else alternative``` would make things much more verbose. Not sure which style is preferred but using the if statement each time is clearer to me.

**Comment by vcunat** [comment]
Yes, that failure gets fixed when I revert your commit.  That's on master.  Curiously, here in this PR the build does not get broken (I tested that yesterday).  I'll try to fix master later today (in ~12h).

**Comment by vcunat** [comment]
For reference, the Linux problem on master should've been solved by #40438.

**Pull request merged**